data-prepper-benchmarks/otel-trace-raw-benchmarks/src/jmh/java/com/amazon/dataprepper/benchmarks/prepper/OtelTraceRawPrepperBenchmarks.java

```diff
@@ -0,0 +1,191 @@
+import io.opentelemetry.proto.trace.v1.Span;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.OperationsPerInvocation;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+
+public class OtelTraceRawPrepperBenchmarks {
+
+    private static final int OPERATIONS_PER_INVOCATION = 10;
+
+    @State(Scope.Benchmark)
+    public static class OtelTraceRawPrepperState {
+        private OTelTraceRawPrepper otelTraceRawPrepper;
+
+        @Param(value = "4")
+        private int concurrencyScale;
+
+        @Setup(Level.Trial)
+        public void setupOtelTraceRawPrepper() {
+            PluginSetting pluginSetting = new PluginSetting("otel_trace_raw_prepper", new HashMap<>()) {{
+                setPipelineName("test-pipeline");
+                setProcessWorkers(concurrencyScale);
+            }};
+            otelTraceRawPrepper = new OTelTraceRawPrepper(pluginSetting);
+        }
+
+        @TearDown(Level.Trial)
+        public void shutdownOtelTraceRawPrepper() {
+            otelTraceRawPrepper.shutdown();
+        }
+    }
+
+    @State(Scope.Thread)
+    public static class TestDataState {
+        private List<byte[]> traceIds = new ArrayList<>();
+        private Map<byte[], byte[]> traceIdToRootSpanId = new HashMap<>();
+        private Map<byte[], List<byte[]>> traceIdToSpanIds = new HashMap<>();
+        private static final Random RANDOM = new Random();
+        private static final List<String> serviceNames = Arrays.asList("FRONTEND", "BACKEND", "PAYMENT", "CHECKOUT", "DATABASE");
+        private static final List<String> traceGroups = Arrays.asList("tg1", "tg2", "tg3", "tg4", "tg5", "tg6", "tg7", "tg8", "tg9");
+        private List<Record<ExportTraceServiceRequest>> bulk;
+
+        @Param(value = "10")
+        protected int batchSize;
+
+        /**
+         * Gets a new trace id. 10% of the time it will generate a new id, and otherwise will pick a random
+         * trace id that has already been generated
+         */
+        private byte[] getTraceId() {
+            if(RANDOM.nextInt(100) < 10 || traceIds.isEmpty()) {
+                final byte[] traceId = getRandomBytes(16);
+                traceIds.add(traceId);
+                return traceId;
+            } else {
+                return traceIds.get(RANDOM.nextInt(traceIds.size()));
+            }
+        }
+
+        /**
+         * Gets a span id and adds to the list of existing span ids
+         */
+        private byte[] getSpanId(final byte[] traceId) {
```

Comment:
Also wondering if it matters that the data isn't super close to what we'd expect in a prod environment. Specifically spans being received in reverse order (parent span first, most recent children last). Maybe "good enough" for this usecase though.

Reply:
For testing this prepper, I need to guarantee that only a single root spanId exists for a particular traceId. Other than that, I do not think closeness to prod matters here.